tests/unit/lib/charms/sdcore_amf/v0/test_fiveg_n2_requirer_interface.py

```diff
@@ -0,0 +1,212 @@
+            relation_name=self.relation_name, remote_app=remote_app_name
+        )
+        self.harness.add_relation_unit(
+            relation_id=relation_id, remote_unit_name=f"{remote_app_name}/0"
+        )
+
+        return relation_id
+
+    @patch.object(DummyFivegN2Requires, "_on_n2_information_available")
+    def test_given_n2_information_in_relation_data_when_relation_changed_then_n2_information_available_event_emitted(  # noqa: E501
+        self,
+        patch_on_n2_information_available,
+    ):
+        relation_id = self._create_relation(remote_app_name=self.remote_app_name)
+
+        relation_data = {
+            "amf_ip_address": "192.168.70.132",
+            "amf_hostname": "amf",
+            "amf_port": "38412",
+        }
+        self.harness.update_relation_data(
+            relation_id=relation_id, app_or_unit=self.remote_app_name, key_values=relation_data
+        )
+
+        patch_on_n2_information_available.assert_called()
+
+    @patch.object(DummyFivegN2Requires, "_on_n2_information_available")
+    def test_given_n2_information_not_in_relation_data_when_relation_changed_then_n2_information_available_event_is_not_emitted(  # noqa: E501
+        self,
+        patch_on_n2_information_available,
+    ):
+        relation_id = self._create_relation(remote_app_name=self.remote_app_name)
+
+        relation_data = {}
+        self.harness.update_relation_data(
+            relation_id=relation_id, app_or_unit=self.remote_app_name, key_values=relation_data
+        )
+
+        patch_on_n2_information_available.assert_not_called()
+
+    @patch.object(DummyFivegN2Requires, "_on_n2_information_available")
+    def test_given_invalid_n2_information_in_relation_data_when_relation_changed_then_n2_information_available_event_is_not_emitted(  # noqa: E501
+        self,
+        patch_on_n2_information_available,
+    ):
+        relation_id = self._create_relation(remote_app_name=self.remote_app_name)
+
+        relation_data = {
+            "amf_ip_address": "192.168.70.132",
+            "amf_hostname": "amf",
+            "amf_port": "invalid_port123",
+        }
+        self.harness.update_relation_data(
+            relation_id=relation_id, app_or_unit=self.remote_app_name, key_values=relation_data
+        )
+
+        patch_on_n2_information_available.assert_not_called()
+
+    def test_given_n2_information_in_relation_data_when_get_amf_ip_address_is_called_then_ip_is_returned(  # noqa: E501
+        self,
+    ):
+        relation_id = self._create_relation(remote_app_name=self.remote_app_name)
+
+        relation_data = {
+            "amf_ip_address": "192.168.70.132",
+            "amf_hostname": "amf",
+            "amf_port": "38412",
+        }
+        self.harness.update_relation_data(
+            relation_id=relation_id, app_or_unit=self.remote_app_name, key_values=relation_data
+        )
+
+        amf_ip_address = self.harness.charm.n2_requirer.amf_ip_address
```

Comment:
I think the instance of the `N2Requires` is not meant to be used that way, is it? As indicated in the comment above, we should be taking params from the event data. 
For testing purposes, you could be checking for certain messages being logged (because the callback function in the dummy charm logs the values of the AMF params). 
WDYT?

Reply:
I agree that the tests were designed wrong, I updated them, now they test that the correct event was actually emitted as expected (or not emitted).

On the other hand the information passed by the provider needs to be available for the requirer at other times than when the event is emitted (for example to populate the config file, that only can be done when all the bits needed are there.)
The logic of course could be pushed to the side of the requirer to get the information from the event and store it on its own to the moment it can use it to populate the config file. (so the provider passes it in the event and that's all).

Other two alternatives are:

1. Having a `fetch_relation_data` function as in the [`data_interfaces` lib.](https://github.com/canonical/data-platform-libs/blob/main/lib/charms/data_platform_libs/v0/data_interfaces.py#L507).

2. Having properties to access these data bits. Like we have in the [`fiveg_nrf` lib](https://github.com/canonical/sdcore-nrf-operator/blob/main/lib/charms/sdcore_nrf/v0/fiveg_nrf.py#L213).

I don't see a clear disadvantage of the current approach, but I'm happy to adapt it to any other design we find to be better.

Reply:
I'm not saying it has disadvantages. I just think it's more complicated than it needs to be. Here's some points to consider:
1. We're introducing additional properties (which btw could use setters if we wanna use them like that), which we don't necessarily need.
2. `n2_available` event sends event data anyway. Do we need redundancy with the properties?
3. We typically use `event.defer` on the requirer side to make sure all the bits are in place before we start real work on the workload.

Reply:
I think it's a good feature to allow the charm to get the relation data in a clean way other than just having it in the event. 
Here's the scenario that makes me think this way, and let me know if it makes sense:

1. Provider sets the data in the relation databag, requirer reacts to `relation_changed` and triggers an event with the data, let the event be called `information_available_event` and its type be `RequirerEvent`.
2. The charm implementing the requirer side will get the data from the event with `event.some_data`, but for that we need to be sure that the event is of type `RequirerEvent` meaning that we need a separate handler for this event, within this handler we can't defer other events and we don't know if all required data bits to populate the config file are available, therefore we need to store the data. If we have a common handler that gathers the bits of data and calls the function to populate the config, we will need to check the type of the event before we call `event.some_data`.
3. We populate the config with the data.

On the other hand if we had a way to fetch the data directly we fetch it on the time of populating the config, if the data is not available we defer the event no matter what is its type.

But again, I remember when working on the `grafana-auth` lib, that a lib should only be triggering events and communicating through events, it shouldn't be raising exceptions or allowing direct access, thus what you are saying would be the way to go in that case.

What do you think? @sanchezfdezjavier @ghislainbourgeois @gruyaume

Reply:
Funny that you mentioned working on the `grafana-auth`, because I was going to bring up my experiences from developing interfaces for Observability as well :D And yes, what you are proposing here, probably goes against the convention we both learnt. I'm not saying we can't go against the convention, but I think we should have a good reason if we do.
As for the scenario - I don't think we need a separate handler. In fact, in most of our charms (and probably most charms in general) we usually have one main handler doing all the heavy lifting. The key concept we leverage is deferring the events which we're not ready to handle. Then such events (along with their event data) go to the queue and the framework re-triggers them whenever any new event happens. Eventually, after XX iterations, all required data is in place and we're done.

Reply:
I collected some examples of libraries that have a mechanism of accessing the relation data other than passing it in the event. Let me know what you think.
[kratos-endpoints](https://github.com/canonical/kratos-operator/blob/main/lib/charms/kratos/v0/kratos_endpoints.py#L136)
[prometheus_scrape](https://github.com/canonical/prometheus-k8s-operator/blob/main/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L1158)
[cos-agent](https://github.com/canonical/grafana-agent-k8s-operator/blob/main/lib/charms/grafana_agent/v0/cos_agent.py#L646)
[hydra-endpoints](https://github.com/canonical/hydra-operator/blob/main/lib/charms/hydra/v0/hydra_endpoints.py#L136)

Reply:
After discussion, agreed to keep it as is.